File: core/src/main/kotlin/org/bitlap/core/sql/rel/BitlapTableScan.kt
package org.bitlap.core.sql.rel

import org.apache.calcite.plan.RelOptCluster
import org.apache.calcite.plan.RelOptTable
import org.apache.calcite.plan.RelTraitSet
import org.apache.calcite.prepare.RelOptTableImpl
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.RelWriter
import org.apache.calcite.rel.core.TableScan
import org.apache.calcite.rel.hint.RelHint
import org.bitlap.core.sql.table.BitlapSqlQueryTable

/**
 * Table scan logical plan, see [org.apache.calcite.rel.logical.LogicalTableScan]
 */
open class BitlapTableScan(
    cluster: RelOptCluster,
    traitSet: RelTraitSet,
    hints: List<RelHint>,
    table: RelOptTable?,
    override var parent: RelNode? = null,
) : TableScan(cluster, traitSet, hints, table), BitlapNode {

    open var converted: Boolean = false

    fun getOTable(): BitlapSqlQueryTable {
        return (this.table as RelOptTableImpl).table() as BitlapSqlQueryTable
    }

    override fun explainTerms(pw: RelWriter): RelWriter {
        return super.explainTerms(pw)
            .item("class", getOTable()::class.java.simpleName)
    }

    override fun withHints(hintList: MutableList<RelHint>): RelNode {
        return BitlapTableScan(cluster, traitSet, hintList, table, parent)
    }

    open fun withTable(table: RelOptTable): BitlapTableScan {
        return BitlapTableScan(cluster, traitSet, hints, table, parent)
    }

    override fun copy(traitSet: RelTraitSet?, inputs: MutableList<RelNode>?): RelNode {
        return this
    }
}

File: bitlap-core/src/main/scala/org/bitlap/core/sql/rel/BitlapTableScan.scala
/*
 * Copyright 2020-2023 IceMimosa, jxnu-liguobin and the Bitlap Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.bitlap.core.sql.rel

import java.util.List as JList

import org.bitlap.core.sql.table.BitlapSqlQueryTable

import org.apache.calcite.plan.RelOptCluster
import org.apache.calcite.plan.RelOptTable
import org.apache.calcite.plan.RelTraitSet
import org.apache.calcite.prepare.RelOptTableImpl
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.RelWriter
import org.apache.calcite.rel.core.TableScan
import org.apache.calcite.rel.hint.RelHint

/** Table scan logical plan, see [org.apache.calcite.rel.logical.LogicalTableScan]
 */
class BitlapTableScan(
  cluster: RelOptCluster,
  traitSet: RelTraitSet,
  hints: JList[RelHint],
  table: RelOptTable,
  var parent: RelNode = null)
    extends TableScan(cluster, traitSet, hints, table),
      BitlapNode {

  var converted: Boolean = false

  def getOTable: BitlapSqlQueryTable = {
    this.table.asInstanceOf[RelOptTableImpl].table().asInstanceOf[BitlapSqlQueryTable]
  }

  override def explainTerms(pw: RelWriter): RelWriter = {
    super
      .explainTerms(pw)
      .item("class", getOTable.getClass.getSimpleName)
  }

  override def withHints(hintList: JList[RelHint]): RelNode = {
    BitlapTableScan(getCluster, getTraitSet, hintList, getTable, parent)
  }

  def withTable(oTable: RelOptTable): BitlapTableScan = {
    BitlapTableScan(getCluster, getTraitSet, getHints, oTable, parent)
  }

  override def copy(traitSet: RelTraitSet, inputs: JList[RelNode]): RelNode = {
    this
  }
}
